# differential_exp_with_R
An example R workflow analyzing data from an experiment in which larval stickleback fish from two populations were exposed to one of two treatments: a conventional microbiota (“CV”) or germ-free (“GF”). RNA was isolated from the intestinal tracts of the experimental fish and used to make individual TruSeq mRNA-seq libraries. The workflow looks at genes differentially expressed between fish of different populations and genesdifferentially expressed between fish exposed or unexposed to microbes.
